Implement builtin map

S-YOU opened the pull request:
Implementing builtin map
Note that variable builtinMap is taken, so I am using builtinMapFn.

trotterdylan commented on runtime/builtin_types.go:
```diff
@@ -193,6 +193,26 @@ func builtinAbs(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
 	return Abs(f, args[0])
 }
 
+func builtinMapFn(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
```
`map()` can take multiple sequences and then supply them all as args to the function:

```
map(...)
    map(function, sequence[, sequence, ...]) -> list

    Return a list of the results of applying the function to the items of
    the argument sequence(s).  If more than one sequence is given, the
    function is called with an argument list consisting of the corresponding
    item of each sequence, substituting None for missing values when not all
    sequences have the same length.  If the function is None, return a list of
    the items of the sequence (or a list of tuples if more than one sequence).
```

S-YOU replied:
Oh, I will work again. Never knew that I can pass many sequence to map.

S-YOU commented:
Added support for multiple sequences, but I can't seem to figure two coverage on
```
if !raised.isInstance(StopIterationType) {
     return nil, raised
}
```
and 
```
// Replace nil with None
if len(list) > 1 {
	lastIndex := len(list) - 1
	for list[lastIndex] == nil {
		list[lastIndex] = None
		lastIndex--
	}
}
```

S-YOU commented:
Yeah, Indeed. I will move zip implemention as separate function.

S-YOU commented:
I am now using similar approach as zip implementation, but I can't able to find way to share code yet, because there is little different on innermost loop.

trotterdylan commented on runtime/builtin_types.go:
```diff
@@ -193,6 +193,81 @@ func builtinAbs(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+				lastIndex--
+			}
+		}
+		// NOTE: append does the capacity doubling logic as needed
```
Remove

trotterdylan commented on runtime/builtin_types.go:
```diff
@@ -193,6 +193,81 @@ func builtinAbs(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+	return result, nil
+}
+
+func builtinMapFn(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
```
Call this `builtinMap` to match the pattern used by other builtin functions.

S-YOU replied:
I cannot, builtinMap is this stuff
```
builtinMap := map[string]*Object{
		"__frame__":      newBuiltinFunction("__frame__", builtinFrame).ToObject(),
		"abs":            newBuiltinFunction("abs", builtinAbs).ToObject(),
		"all":            newBuiltinFunction("all", builtinAll).ToObject(),
```

trotterdylan commented on runtime/builtin_types.go:
```diff
@@ -193,6 +193,81 @@ func builtinAbs(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
 	return Abs(f, args[0])
 }
 
+func zipLongestImpl(f *Frame, args Args) ([]*Object, *BaseException) {
```
Call this zipLongest since it's slightly simpler. Also, please move it to the end of the file. I try to group all the helper functions at the end of the file (although I'm sure I've missed some).

S-YOU replied:
Sure

trotterdylan commented on runtime/builtin_types.go:
```diff
@@ -193,6 +193,81 @@ func builtinAbs(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+			return nil, raised
+		}
+		iters[i] = iter
+	}
```
Seems like you could factor this iterator collecting logic out of map() and zip() into a function that returns []*Object. But that could wait until a later PR if you prefer.

S-YOU replied:
Let me do this after both PR are merged, because switching between two branch to refactor or through web take time.

trotterdylan replied:
Now that this is merged with the zip() change, would you mind updating builtinZip() to use initIters() as well?

S-YOU replied:
Sure

trotterdylan commented on runtime/builtin_types.go:
```diff
@@ -193,6 +193,81 @@ func builtinAbs(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+			elem, raised := Next(f, iter)
+			if raised != nil {
+				if raised.isInstance(StopIterationType) {
+					continue
```
As in zip(), you need to RestoreExc(nil, nil)

trotterdylan commented on runtime/builtin_types.go:
```diff
@@ -193,6 +193,81 @@ func builtinAbs(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+			break
+		}
+		// Replace nil with None
+		if len(elems) > 1 {
```
Could you set `elems[i] = None` when you detect StopIteration above?

S-YOU replied:
I thought nil where more than one, looks like that was not the case. 
Much simpler now. Thank you.

trotterdylan commented on runtime/builtin_types.go:
```diff
@@ -193,6 +193,81 @@ func builtinAbs(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+		noItems := true
+		elems := make([]*Object, argc)
+		for i, iter := range iters {
+			elem, raised := Next(f, iter)
```
Raising an exception requires allocating a new object which is relatively expensive. In general it's good to avoid as many allocations as possible. So rather than calling Next() on exhausted iterators, it would probably be cheaper to set `iters[i] = nil` and check for nil before calling Next().

trotterdylan commented on runtime/builtin_types.go:
```diff
@@ -193,6 +193,81 @@ func builtinAbs(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+		return nil, raised
+	}
+	for _, tuple := range z {
+		if pred.Type() == NoneType {
```
`if pred == None {`

trotterdylan commented on runtime/builtin_types.go:
```diff
@@ -193,6 +193,81 @@ func builtinAbs(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+				result = append(result, tuple)
+			}
+		} else {
+			ret, raised := pred.Call(f, toTupleUnsafe(tuple).elems, nil)
```
Rather than allocating a Tuple and then not using it in most cases, maybe zipLongest should return `[][]*Object.`

S-YOU commented:
Fixed all, except renaming builtinMapFn to builtinMap, because builtinMap is in used by

```
builtinMap := map[string]*Object{
		"__frame__":      newBuiltinFunction("__frame__", builtinFrame).ToObject(),
		"abs":            newBuiltinFunction("abs", builtinAbs).ToObject(),
		"all":            newBuiltinFunction("all", builtinAll).ToObject(),
```

also added `initIters` function, to collect iterators, let me know if you have better name.

S-YOU commented:
Rebased and updated to use initIters.

trotterdylan commented:
This is great! Thanks for doing this.
